cms/core/blocks/related.py

```diff
@@ -1,44 +1,81 @@
 from typing import TYPE_CHECKING, Any
 
+from django.forms import ValidationError
 from django.utils.text import slugify
-from django.utils.translation import gettext as _
-from wagtail.blocks import CharBlock, ListBlock
+from django.utils.translation import gettext_lazy as _
+from wagtail.blocks import CharBlock, ChoiceBlock, DateBlock, ListBlock, StructBlockValidationError
 
-from .base import LinkBlock
+from cms.core.blocks.base import LinkBlock
+from cms.core.enums import RelatedContentType
 
 if TYPE_CHECKING:
+    from django_stubs_ext import StrOrPromise
     from wagtail.blocks.list_block import ListValue
 
+    from cms.core.blocks.base import LinkBlockStructValue
 
-class RelatedContentBlock(LinkBlock):
-    """Related content block with page or link validation."""
 
+class LinkBlockWithDescription(LinkBlock):
     description = CharBlock(required=False)
 
 
+class RelatedContentBlock(LinkBlockWithDescription):
+    """Related content block with page or link validation."""
+
+    content_type = ChoiceBlock(
+        choices=RelatedContentType.choices,
+        help_text="Select the type of related content.",
+        required=False,
+    )
+    release_date = DateBlock(required=False)
+
+    def clean(self, value: "LinkBlockStructValue") -> "LinkBlockStructValue":
+        """Validate the related content based on its type."""
+        cleaned_value = super().clean(value)
+
+        errors = {}
+
+        if cleaned_value["external_url"] and not cleaned_value["content_type"]:
+            errors["content_type"] = ValidationError("You must select a content type when providing an external URL.")
+
+        if errors:
+            raise StructBlockValidationError(block_errors=errors)
+
+        return cleaned_value
+
+
 class RelatedLinksBlock(ListBlock):
     """Defines a list of links block."""
 
-    def __init__(self, search_index: bool = True, **kwargs: Any) -> None:
+    def __init__(self, search_index: bool = True, add_heading: bool = False, **kwargs: Any) -> None:
+        """Related links block.
+
+        Args:
+            search_index (bool, optional): Whether to index the block for search.
+            add_heading (bool, optional): Whether to add a "Related links" heading to the block.
+            **kwargs: Additional keyword arguments to pass to the ListBlock constructor.
+        """
         super().__init__(RelatedContentBlock, search_index=search_index, **kwargs)
 
-        self.heading = _("Related links")
-        self.slug = slugify(self.heading)
+        self.heading = _("Related links") if add_heading else ""
+        self.slug = slugify(self.heading) if add_heading else ""
```

Comment:
Q: Could someone please refresh my memory on what's the role of the `slug` here. It's later passed to as an `id` of a `<section>` in the block HTML template, but then what? 🧐

Reply:
It's for the table of contents, so that we have anchor points (`#slug`)